metricbeat/module/system/socket_summary/sockstat_linux.go

```diff
@@ -0,0 +1,125 @@
+// Licensed to Elasticsearch B.V. under one or more contributor
+// license agreements. See the NOTICE file distributed with
+// this work for additional information regarding copyright
+// ownership. Elasticsearch B.V. licenses this file to you under
+// the Apache License, Version 2.0 (the "License"); you may
+// not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing,
+// software distributed under the License is distributed on an
+// "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+// KIND, either express or implied.  See the License for the
+// specific language governing permissions and limitations
+// under the License.
+
+// +build linux
+
+package socket_summary
+
+import (
+	"bufio"
+	"fmt"
+	"os"
+
+	"github.com/pkg/errors"
+
+	"github.com/elastic/beats/libbeat/common"
+)
+
+//SockStat contains data from /proc/net/sockstat
+type SockStat struct {
+	//The count of all sockets in use on the system, in the most liberal definition. See `ss -s` and `ss -a` for more.
+	SocketsUsed int
+	//Total in use TCP sockets
+	TCPInUse int
+	//Total 'orphaned' TCP sockets
+	TCPOrphan int
+	//Sockets in TIME_WAIT
+	TCPTW int
+	//Total allocated sockets
+	TCPAlloc int
+	//Socket memory use, in pages
+	TCPMem int
+	//In use UDP sockets
+	UDPInUse int
+	//Socket memory use, in pages
+	UDPMem int
+	//UDP-Lite in use sockets
+	UDPLiteInUse int
+	//In Use raw sockets
+	RawInUse int
+	//FRAG sockets in use
+	FragInUse int
+	//Frag memory, in bytes
+	FragMemory int
+}
+
+//get a list of platform-specific enhancements and apply them to our mapStr object.
+func applyEnhancements(data common.MapStr, m *MetricSet) (common.MapStr, error) {
+	stat, err := parseSockstat(m.sockstat)
+	if err != nil {
+		return nil, errors.Wrap(err, "error getting sockstat data")
+	}
+	data["tcp"].(common.MapStr)["orphan_sockets"] = stat.TCPOrphan
+	data["tcp"].(common.MapStr)["socket_memory"] = os.Getpagesize() * stat.TCPMem
```

Comment:
This can panic if `data["tcp"]` is not initialized, add some type checks, or use `Put`:
```
data.Put("tcp.orphan_sockets", stat.TCPOrphan)
data.Put("tcp.socket_memory", os.Getpagesize() * stat.TCPMem)
```